### PR TITLE
File-per-class and single file bindings

### PR DESCRIPTION
**Pull request by mahesh-hegde**
Introduces configuration parameter: `output_structure`. It can take 2 values `package_structure` or `single_file`. In single file mode, output path must end with '.dart'.

In `package_structure` mode, output is written with each java class mapped to one dart file. It's a change from previous one where it was file-per-package.

closes: #90 , possibly #89 but we might leave it open for recursive _all.dart

It should ideally have been 2 PRs, but I am in a hurry and also don't want my dev branches diverge too much from upstream which will be a hassle to merge.

### CODE REVIEW

**Comment by mahesh-hegde** [comment]
Wiindooooows :coffee:

File: .github/workflows/test-package.yml [hunk 1]
```diff
@@ -331,11 +331,11 @@ jobs:
       - run: flutter build apk
         working-directory: ./jnigen/example/notification_plugin/example
       - name: re-generate bindings
-        run: flutter pub run jnigen -Doutput.dart.path=_dart -Doutput.c.path=_c --config jnigen.yaml
+        run: flutter pub run jnigen -Doutput.dart.path=_temp.dart -Doutput.c.path=_c --config jnigen.yaml
```
Comment (mahesh-hegde) [inline .github/workflows/test-package.yml:334]: (This example was changed to generate single dart file)

**Review by dcharkes — APPROVED** [review]
lgtm!

Works perfectly in my own test project. 🚀

File: jnigen/tool/pre_commit_checks.dart [hunk 2]
```diff
@@ -174,20 +174,20 @@ void main() async {
...
   final comparePdfboxBindings = Runner(
       "Generate & compare PdfBox Bindings", "jnigen/example/pdfbox_plugin")
     ..chainCommand("dart", [
       "run",
       "jnigen",
       "--config",
       "jnigen.yaml",
-      "-Doutput.c.path=src_temp",
-      "-Doutput.dart.path=lib_temp",
+      "-Doutput.c.path=src_temp/",
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jnigen/tool/pre_commit_checks.dart:189]: Actually I wrote `Uri.directory(src.toFilePath())` to ensure a directory URI is created either way.

But I can change that to error instead if the path doesn't end with '/'.

What's better?

File: jnigen/example/notification_plugin/lib/notifications.dart [hunk 3]
```diff
@@ -15,7 +17,10 @@ import "dart:ffi" as ffi;
 import "package:jni/internal_helpers_for_jnigen.dart";
 import "package:jni/jni.dart" as jni;
 
-import "../../_init.dart" show jniLookup;
+// Auto-generated initialization code.
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jnigen/example/notification_plugin/lib/notifications.dart:20]: Now we have _package.dart in multi file mode, should we rename _init.dart to something more hideous?

File: jnigen/tool/pre_commit_checks.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (dcharkes) [reply jnigen/tool/pre_commit_checks.dart:189]: Well, it's okay if it works. However, maybe add a warning to educate our users they should end directories with a directory separator. (And just continue working after the warning.)

[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jnigen/tool/pre_commit_checks.dart:189]: I would prefer an error over a warning. Because people always ignore warnings.

File: jnigen/example/notification_plugin/lib/notifications.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (dcharkes) [reply jnigen/example/notification_plugin/lib/notifications.dart:20]: It's probably fine.

File: jnigen/tool/pre_commit_checks.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (dcharkes) [reply jnigen/tool/pre_commit_checks.dart:189]: Okay, I don't mind an error. People will run into it when configuring their yaml in the beginning immediately anyways.

File: jnigen/lib/src/config/config.dart [hunk 4]
```diff
@@ -133,6 +133,13 @@ T _getEnumValueFromString<T>(
   return value;
 }
 
+void _ensureIsDirectory(String name, Uri path) {
+  if (!path.toFilePath().endsWith(Platform.pathSeparator)) {
+    throw ArgumentError('$name must be a directory path. If using YAML '
```
Comment (dcharkes) [inline jnigen/lib/src/config/config.dart:138]: Do we need test cases that expect an error to be thrown for the negative cases we're trying to prevent? Possibly also for all other kinds of configs which we give error on. So this probably should be a separate PR then. This would increase our code coverage.

Comment (mahesh-hegde) [reply jnigen/lib/src/config/config.dart:138]: Yeah I think error paths aren't checked well in config, binding generators. Last time I checked lcov report for support library I was left with same impression.


Also some checks like pdfbox plugin test aren't covered by test coverage.

There could be separate PR for test coverage.

Ideally should also have test coverage for summarizer and C code in support library. Not sure how to combine all results using coveralls.

Comment (dcharkes) [reply jnigen/lib/src/config/config.dart:138]: Let's do it in a separate pr, I've filed https://github.com/dart-lang/jnigen/issues/101.

**Comment by mahesh-hegde** [comment]
Didn't update the CI file itself, and Uri.directory() when reading C meant it wouldn't catch the same error in c.output.path. :see_no_evil:
